rules/no-array-fill-with-reference-type.js

```diff
@@ -0,0 +1,535 @@
+// @ts-check
+import {findVariable} from '@eslint-community/eslint-utils';
+import {
+	isCallExpression,
+	isFunction,
+	isMemberExpression,
+	isMethodCall,
+	isNewExpression,
+	isRegexLiteral,
+} from './ast/index.js';
+
+// @ts-check
+const MESSAGE_ID_ERROR = 'no-array-fill-with-reference-type/error';
+const messages = {
+	[MESSAGE_ID_ERROR]:
+		'Avoid using `{{actual}}` with reference type{{type}}. Use `Array.from({ ... }, () => { return independent instance })` instead to ensure no reference shared.',
+};
+
+const DEFAULTS = {
+	// Not check for function expressions by default because it is rare to fill an array with a function and add properties to it.
+	allowFunctions: true,
+	// The same reason as above.
+	allowRegularExpressions: true,
+};
+
+const RECURSION_LIMIT = 5;
+
+/** @param {import('eslint').Rule.RuleContext} context */
+const create = context => ({
+	/**
+	  @param {import('estree').CallExpression} node
+	 */
+	CallExpression(node) {
+		// Check all `fill` method call even if the object is not Array because we don't know if its runtime type is array
+		// `arr.fill()` or `new Array().fill()` or `Array.from().fill()`
+		const isArrayFill = isMethodCall(node, {method: 'fill', argumentsLength: 1});
+
+		// `Array.from().map` or `Array.from(arrayLike, mapper)`
+		const isArrayFrom = isMethodCall(node, {object: 'Array', method: 'from'});
+
+		if (!isArrayFill && !isArrayFrom) {
+			return;
+		}
+
+		const fillArgument = isArrayFill
+			? node.arguments[0]
+			: getReturnNodeOfArrayDotFrom(node, context);
+
+		const [is, resolvedNode] = isReferenceType(fillArgument, context);
+
+		if (!is) {
+			return;
+		}
+
+		const actual = isMethodCall(
+			// @ts-expect-error
+			node.callee.object,
+			{object: 'Array', method: 'from'},
+		)
+			? 'Array.from().fill()'
+			: 'Array.fill()';
+
+		const type = getType(resolvedNode, context);
+
+		return {
+			node: fillArgument,
+			messageId: MESSAGE_ID_ERROR,
+			data: {
+				actual,
+				type: type ? ` (${type})` : '',
+			},
+		};
+	},
+});
+
+/**
+
+ @param {import('estree').CallExpression} functionNode
+ @param {RuleContext} context
+ @returns
+ */
+function getReturnNodeOfArrayDotFrom(functionNode, context) {
+	const secondArgument = functionNode.arguments[1];
+
+	// Array.from({ length: 10 }, () => { return sharedObject; });
+	let result;
+	if (secondArgument && isFunction(secondArgument)) {
+		result = getReturnIdentifier(secondArgument, context);
+		// @ts-expect-error node always has a parent
+	} else if (isMemberExpression(functionNode.parent, 'map')) {
+		// Array.from({ length: 10 }).map(() => { return sharedObject; });
+		// @ts-expect-error node always has a parent
+		result = getReturnIdentifier(functionNode.parent.parent.arguments[0], context);
+	}
```

Comment:
This part should be done in the `CallExpression`, and pass the callback function to this function.